community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/LogicalPlanConverter.scala

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright (c) 2002-2017 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.compatibility.v3_3
+
+import java.lang.reflect.Constructor
+
+import org.neo4j.cypher.internal.compatibility.v3_3.SemanticTableConverter.ExpressionMapping3To4
+import org.neo4j.cypher.internal.frontend.v3_3.ast.{Expression => ExpressionV3_3}
+import org.neo4j.cypher.internal.frontend.v3_3.{SemanticDirection => SemanticDirectionV3_3, ast => astV3_3, symbols => symbolsV3_3, InputPosition => InputPositionV3_3}
+import org.neo4j.cypher.internal.frontend.{v3_3 => frontendV3_3}
+import org.neo4j.cypher.internal.ir.v3_3.{IdName => IdNameV3_3}
+import org.neo4j.cypher.internal.ir.v3_4.{PlannerQuery, IdName => IdNameV3_4}
+import org.neo4j.cypher.internal.ir.{v3_3 => irV3_3, v3_4 => irV3_4}
+import org.neo4j.cypher.internal.util.v3_4.Rewritable.{DuplicatableProduct, RewritableAny}
+import org.neo4j.cypher.internal.util.v3_4.{symbols => symbolsV3_4, _}
+import org.neo4j.cypher.internal.util.{v3_4 => utilV3_4}
+import org.neo4j.cypher.internal.v3_3.logical.plans.{LogicalPlan => LogicalPlanV3_3}
+import org.neo4j.cypher.internal.v3_3.logical.{plans => plansV3_3}
+import org.neo4j.cypher.internal.v3_4.expressions.{Expression => ExpressionV3_4}
+import org.neo4j.cypher.internal.v3_4.logical.plans.{LogicalPlan => LogicalPlanV3_4}
+import org.neo4j.cypher.internal.v3_4.logical.{plans => plansV3_4}
+import org.neo4j.cypher.internal.v3_4.{expressions => expressionsV3_4}
+import org.neo4j.cypher.internal.compiler.{v3_3 => compilerV3_3}
+
+import scala.collection.mutable
+import scala.collection.mutable.{HashMap => MutableHashMap}
+import scala.util.{Failure, Success, Try}
+
+object LogicalPlanConverter {
```

Comment:
In general, the `.instanceOf[]` calls make the big match clause difficult to read. I would like to see the children cast into variables that we can then use to create the new plan. Example:
```
        case (plan: plansV3_3.TriadicSelection, children: Seq[AnyRef]) =>
          plansV3_4.TriadicSelection(children(1).asInstanceOf[LogicalPlanV3_4],
            children(5).asInstanceOf[LogicalPlanV3_4],
            children(0).asInstanceOf[Boolean],
            children(2).asInstanceOf[IdNameV3_4],
            children(3).asInstanceOf[IdNameV3_4],
            children(4).asInstanceOf[IdNameV3_4])(new PlannerQueryWrapper(plan.solved))
``` 

could instead be:

```
        case (plan: plansV3_3.TriadicSelection, children: Seq[AnyRef]) =>
          val lhs = children(1).asInstanceOf[LogicalPlanV3_4]
          val rhs = children(5).asInstanceOf[LogicalPlanV3_4]
          val positivePredicate = children(0).asInstanceOf[Boolean]
          val sourceId = children(2).asInstanceOf[IdNameV3_4]
          val seenId = children(3).asInstanceOf[IdNameV3_4]
          val targetId = children(4).asInstanceOf[IdNameV3_4]

          plansV3_4.TriadicSelection(lhs, rhs, positivePredicate, sourceId, seenId, targetId)(new PlannerQueryWrapper(plan.solved))
```

Reply:
Would naming the arguments work for you too? Like

```
plansV3_4.TriadicSelection(lhs = children(1).asInstanceOf[LogicalPlanV3_4],
            rhs = children(5).asInstanceOf[LogicalPlanV3_4],
            positivePredicate = children(0).asInstanceOf[Boolean],
            sourceId = children(2).asInstanceOf[IdNameV3_4],
            seenId = children(3).asInstanceOf[IdNameV3_4],
            targetId = children(4).asInstanceOf[IdNameV3_4])
```